### DISCUSSION TITLE
Model custom db connection is reset to default during first initialization

### DISCUSSION THREAD

**Discussion by jonTravens** [livewire/livewire#5155] [7. Bugs (archived)] [answered]
In my app I have the need to use different DB Connection settings on certain models.

So I set the protected `$connection` property on the relevant models to the db connection I needed.

Using Tinker I could verify that it works and new changes are saved using the specificied db connection , but once I tested in a Livewire component, I realized that the `$connection` property is reset to my default DB connection setting.

To illustrate, here are bits of the code I use.

In `config\database` default is set to `mysql`, and I also create another connection name `foo`
```php 
// config/database.php

 'default' => env('DB_CONNECTION', 'mysql'),

 'connections' => [

        'mysql' => [
            ...
        ],
        
        'foo' => [
          ...
        ]
  ]

```

In my model I set the connection property to  `foo`

```php
<?php
namespace App\Models;

class Section extends Model
{


    protected $connection = 'foo';

    ...

}
?>
```

The Livewire component looks like this
``` php
<?php

namespace App\Http\Livewire;

use App\Models\Section;

class EditSection extends Component
{

    public Section $section;

    public function mount(Section $section)
    {
        $this->section = $section;

       // At this point, $this->section->getConnectionName() returns `foo`
    }


    public function save()
    {  
        // At this point, $this->section->getConnectionName() returns `mysql` (default)
    }

?>
```

I've tried to understand at what point the connection name was reset to default, so I started digging in Livewire source (what a ride !! 😅)

As far as I could go, the value is changed within a method that outputs the html when rendering the view.

This happens right after this line https://github.com/livewire/livewire/blob/63ad79f9bbad51028e1aaeca774f2f5e22f6578d/src/Component.php#L235

To be clear 
```php
 app('view')->share('errors', $errors);
 app('view')->share('_instance', $this);

 // At this point, the model connection name returns `foo`

 $output = $view->render();

 // At this point, the model connection name returns `mysql`

 app('view')->share('errors', $previouslySharedErrors);
```

I could even trace it a little bit further inside the RendersLivewireComponents trait used in LivewireViewCompilerEngine as a replacement to the default Compiler engine.
 https://github.com/livewire/livewire/blob/63ad79f9bbad51028e1aaeca774f2f5e22f6578d/src/ComponentConcerns/RendersLivewireComponents.php#L49

At this point I'm not even sure it is a Livewire issue or a Laravel issue.
Yet I'm pretty sure it's a bug, but I couldn't get any further and thought it might be better to ask for help 🤗

I can work around this by setting the connection before saving, but I would be better that the connection set on the model is preserved. 

Thanks in advance for helping me solve this.

**Comment by Martleby** [comment, marked as answer]
Hello!

I have multiple connections in my application. I just wrote a quick component to test this and I don't get the same behaviour you're seeing. My connection string stays put to what I set it to in the model. 

Are you doing anything in the model that does something to connection strings? Are you using any traits on the model that might interfere or do you have any middlewares/other code that might be doing something? 

Sorry I can't be much more help than that, but based on my test I don't think Livewire is doing this.

Cheers,
Martin.

**Reply by jonTravens** [reply]
Hi Martin,

Thanks for taking some time to take a look at this !

I do have some other traits on this model but on the part I managed to debug I didn’t stubble on any code linked to these.
The traits used are not related to the connection property as far as I know.  

I’ll make other tests tomorrow though, to make sure this issue has nothing to do with any other code that could interfere with my model during component initialization. 

Cheers,

Jon

**Reply by joshhanley** [reply]
@jonTravens yeah to expand on what @Martleby is saying, Livewire actually keeps track of the database connection a model uses when it serialises them, so should restore all that information happily.

You can actually check this in devtools in your browser. If you open devtools > network tab, and then refresh your page and then make a Livewire request (hit save button) and select the request, you should be able to see the data Livewire is sending to the backend. Under payload, look at `serverMemo > dataMeta > models > section (your property name) > connection` and see what connection name Livewire is sending to the backend and see if it is the correct db connection or not.

If the db connection name is correct, then something else is going on when the model is restored by Livewire.
But if it is wrong, then something is changing it when Livewire serialises the model initially for sending to the front end.

Can you also try making an isolated test and see if the issue is still happening and let us know?

Hope this helps!

**Reply by jonTravens** [reply]
Hi ! 👋

Thanks so much to the both of you for taking some of your time to test my case and give me your input ! It helped a lot!

I wrote some isolated tests that passed, confirming it was most probably not coming from Livewire.
Then I started rebuilding my component bit by bit to understand at what moment the change is done.

**I finally figured it out** 🥳, but honestly I feel pretty stupid because it was indeed - as you both said - something interfering in my code... 😓


The error was due to the configuration of the relationship between 2 models that are not using the same db connection... 
I read (or misread) an article explaining how to set multi connection in Laravel and did this :

```php
class Section extends Model
{

  protected $connection = 'mysql_2';


/**
   * Get the Guide associated with the Section.
   */
public function entries()
{
       // I thought I had to set the connection name for the related model....
      return $this->setConnection('mysql')->hasMany(Entry::class);
}

```

In my component I was retrieving the Entries associated with the Section edited, therefore changing the connection property before Livewire serializes it.... 😅


The good news is, it's not a Livewire or Laravel bug, so both are still awesome 🤗


Thanks again and cheers for now,

Jon

**Reply by joshhanley** [reply]
@jonTravens 🥳 no worries! Glad it helped 🙂

I've not used multiple database connections before, are you able to explain what fixed it? Did you just remove that `setConnection` call from the entries relation and everything started working?

**Reply by jonTravens** [reply]
Hi @joshhanley 

Sure I'm happy to explain 🙂

Actually removing the setConnection call was not enough to make it work

As I understand it know, the trait on the Model that manage relationships (HasRelationship) tries to resolve the connection name for the related model. 
1. It calls getConnectionName() on the related model to check if a connection property was set on the model instance (it defaults to null if not set manually)
2. If null, it reuses the same connection as the model initiating the relationship call. 

So if you want to use at least 2 different DB connection, and want to keep relationship between 2 models using different connection, you will need to set the connection property on both models.

To avoid having to set manually the name of the default DB on every model that use the default connection, I just created this simple trait 

```php
<?php

namespace App\Models\Concerns;

trait UsesDefaultConnection
{
    public function getConnectionName()
    {
        return config('database.default');
    }
}

```

Hope it answers your question, if you want more details don't hesitate to get back to me 😉

**Reply by joshhanley** [reply]
@jonTravens ah yep, make sense! Thanks for explaining 🙂 bit a pain you have to set the connection on models that are using the default connection too. Your trait solution is nice and simple though!

**Reply by jonTravens** [reply]
@joshhanley Yeah at first I thought it made sense that a model with no connection set would use the default, and in a simple use case when you fetch some items in a table that’s the way the connection is actually resolved. 
But for relationships I guess the Laravel team went with a logic that matches the most common use case: two related models usually are in the same database. 
A relationship between two tables in different databases seems like an edge case, or at least an uncommon one 😅
But yeah I guess the simple trait solves this pretty easily.

**Reply by joshhanley** [reply]
@jonTravens yeah that would have been my assumption too, that any models that have nothing specified would use the default (regardless of how they are loaded). Have you tried asking around on the Laravel forums, maybe it's a bug with relationships?

**Reply by joshhanley** [reply]
@jonTravens what is your use-case, if you don't mind me asking? What are the different models and why are they in different databases and what are the dbs?

**Reply by jonTravens** [reply]
I actually didn’t ask on Laravel forums, the logic in the method of the HasRelationship trait is pretty straightforward and seems coded like this on purpose. 

What do you think? 🤔 

```php
/**
     * Create a new model instance for a related model.
     *
     * @param  string  $class
     * @return mixed
     */
    protected function newRelatedInstance($class)
    {
        return tap(new $class, function ($instance) {
            if (! $instance->getConnectionName()) {
                $instance->setConnection($this->connection);
            }
        });
    }
```


In my case the multi db is for an app I’m working on, (https://cozylist.co if you’re curious 😉). 
I am exploring the possibility to use SQLite as a production database.
I read about some devs using it, like Pieter Levels with NomadList, but to overcome the writing limitations they put each table in its own SQLite database. 
Not sure yet if that’s a good idea, but I’m toying with the idea for now 😋

**Reply by Martleby** [reply]
I tell ya, Eloquent is genius!

My application aggregates a lot of information across the business. I'm dealing with various different databases, across multiple servers, and across multiple database engines. So for example the MIS system is SQL Server, as is HR but on a different server, finance is MySql on yet another server, Payments is a law unto itself and yet another server and MySql, then I have my own SQL Server db... yeah, it gets complicated!

Eloquent makes this so straight forward:

Setup your connections.
Set $connection per model.
Setup your relationships.
I also like to set the fully qualified table name.

Job's done! And importantly all of that is abstracted so if I do $user->orders or $user->staffAppraisals  I don't have to worry about the underlying database structures/servers/etc* as Eloquent takes care of that for me. It all appears as "one" database.

*Naturally things get more complicated trying to do cross server, cross database queries - but happily Sql Server has linked servers, plus you can throw Eloquent models over DB views, and I have a few other tricks and techniques I use.

Sometimes I wish I had control of all the databases in question and could merge them all into one... but where would the fun in that be?

Cheers,
Martin.

**Comment by joshhanley** [comment]
@jonTravens ah yeah, does seem pretty intentional. I wonder why though...

Oh nice! Looks good 😁

Ah yep fair enough! Any particular reason why SQLite over other databases? Other than just for fun/interests sake of course 😆

@Martleby oh wow! Thanks for sharing! Yeah that would certainly make life interesting 😁 But man that's a lot of databases to deal with! How do you handle schema changes done by a different department?

Yep Eloquent is great, does make life so much easier! 

TIL, I didn't know you could put models over DB views (I always just use models 😅), do you have any references for how to do that? Would like to have a look.

**Reply by Martleby** [reply]
Well to attempt to head off problems before they emerge, we have a complete test environment that replicates all the live systems we interact with. Then I wrote a lot of feature tests to test things out. This catches a lot of things, and there's an added benefit that those external dbs are all fairly mature and unlikely to change too much now. But there's the occasional, inevitable, issue that gets into production... which we fix when we see them and write more tests to catch them later.... it's always interesting when that happens 😁

Here's how I do models over DB views (it's easy!)

I like to use migrations to manage my views. In my up method I do something like so:

```php
DB::connection('myDb')->statement(
            "CREATE VIEW user_stats AS ( .... )"
);
```

Then all you need do is create a model in the normal way and point it at the view as if it were a table. And that's it!

As far as Eloquent knows this is just another table in the database, and so the usual Eloquent goodness applies - you can join to it, and from it, eager loading works etc.

I find this really powerful for things like statistics that you want generated in the DB (eg. stuff like last login, no. of posts, post to comment ratio that sort of thing) that you want to access like a property on the object like so $user->stats->number_of_likes etc. 

Another really good use case is simplifying a complex set of relations. For instance I have a view that abstracts a query with 10 different joins to get its data, so rather than having to do that in Eloquent with lots of different models or with a query builder  I can just do $user->attendance->attendance_last_7_days. And if I need it, I can eager load multiples.

I can also use similar technique to help make very distant models closer by using the view as part of a hasManyThrough relation.

Probably lots of other uses too, along with a few caveats!

Cheers,
Martin.

**Reply by joshhanley** [reply]
@Martleby ah yep, fair enough! Ha yeah I bet!

Oh sweet! Thanks for sharing, going to have to have a play with that 😁 Yeah I can see a few use cases already for moving some tables I have a heap of joins in to a DB view!
